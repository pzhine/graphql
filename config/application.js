// @flow

import url from 'url';

export default {
  restApiEndpoint: {
    allBookings:
      'https://booking-api.skypicker.com/api/v0.1/users/self/bookings',
    singleBooking: (bookingId: number, simpleToken: string) =>
      queryWithParameters(
        `https://booking-api.skypicker.com/api/v0.1/users/self/bookings/${bookingId}`,
        {
          simple_token: simpleToken,
        },
      ),
    allFlights: (queryParameters: ?Object = null) =>
      queryWithParameters('https://api.skypicker.com/flights', {
        ...queryParameters,
        v: 3,
      }),
    login: 'https://auth.skypicker.com/v1/user.login',
    identity: 'https://auth.skypicker.com/v1/user.get',
    airlines: 'https://api.skypicker.com/airlines?v=2',
    allLocations: (queryParameters: Object) =>
      queryWithParameters(
        'https://api.skypicker.com/locations',
        queryParameters,
      ),
    rates: 'https://api.skypicker.com/rates',
    allFAQ: (search: string) =>
      queryWithParameters(
        'https://api.skypicker.com/knowledgebase/api/v1/search',
        { q: search, autocomplete: true },
      ),
    allFAQCategories: (faqCategoryId: ?number) => {
      const categoryId = faqCategoryId ? `/${faqCategoryId}` : '';

      return `https://api.skypicker.com/knowledgebase/api/v1/categories${categoryId}`;
    },
    FAQArticle: (faqArticleId: string) => {
      return `https://api.skypicker.com/knowledgebase/api/v1/articles/${faqArticleId}`;
    },
    FAQArticleVote: (faqArticleId: string) =>
      `https://api.skypicker.com/knowledgebase/api/v1/articles/${faqArticleId}/vote`,
  },
  auth: {
    basicToken: String(process.env.AUTH_BASIC_TOKEN),
    digest: String(process.env.AUTH_DIGEST),
  },
};

export function queryWithParameters(
  absoluteUrl: string,
  queryParameters: ?Object = null,
) {
  const urlObject = url.parse(absoluteUrl, true);
  // intentional ==, can be null or undefined
  if (queryParameters != null) {
    urlObject.query = filterParameters(queryParameters);
  }
  return url.format(urlObject);
}

function filterParameters(parameters: Object) {
  const filteredParameters = parameters;
  Object.keys(parameters).forEach(key => {
    const val = parameters[key];
    if (val === null || val === undefined || val === '') {
      delete filteredParameters[key];
      return;
    }

    if (Array.isArray(val)) {
      filteredParameters[key] = val.toString();
    }

    if (typeof val === 'string') {
      filteredParameters[key] = val.trim();
    }
  });
  return filteredParameters;
}
