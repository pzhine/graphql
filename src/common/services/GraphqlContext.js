// @flow

import DataLoader from 'dataloader';

import IdentityDataloader from '../../identity/dataloaders/Identity';
import createBookingLoader from '../../booking/dataloaders/Booking';
import createAirlineLoader from '../../flight/dataloaders/Airline';
import createRatesLoader from '../dataloaders/Rates';
import createFAQLoader from '../../FAQ/dataloaders/searchFAQ';
import createFAQCategoryLoader from '../../FAQ/dataloaders/getFAQCategory';
import createFAQArticleLoader from '../../FAQ/dataloaders/getFAQArticle';
import BookingsLoader from '../../booking/dataloaders/Bookings';
import LocationSuggestionsLoader from '../../location/dataloaders/LocationSuggestions';
import LocationLoader from '../../location/dataloaders/Location';
import FlightLoader from '../../flight/dataloaders/Flight';
import OptionsStorage from './context/OptionsStorage';
import HotelsAvailability from '../../hotel/dataloaders/HotelsAvailability';
import { type SearchParameters as HotelKey } from '../../hotel/dataloaders/flow/SearchParameters';
import HotelByID from '../../hotel/dataloaders/HotelByID';
import HotelCities from '../../hotel/dataloaders/HotelCities';
import CitiesByID from '../../hotel/dataloaders/CitiesByID';
import HotelRoomsLoader from '../../hotel/dataloaders/HotelRooms';
import HotelRoomAvailabilityLoader from '../../hotel/dataloaders/HotelRoomAvailability';
import HotelRoomBeddingLoader from '../../hotel/dataloaders/RoomBedding';
import PriceStatsLoader from '../../hotel/dataloaders/PriceStats';

import type { Booking } from '../../booking/Booking';
import type { Airline } from '../../flight/Flight';
import type { HotelType } from '../../hotel/dataloaders/flow/HotelType';
import type { City } from '../../hotel/dataloaders/flow/City';
import type { Args as FAQArgs } from '../../FAQ/dataloaders/searchFAQ';
import type {
  Args as FAQCategoryArgs,
  FAQArticleItem,
} from '../../FAQ/dataloaders/getFAQCategory';
import type {
  Args as FAQCArticleArgs,
  FAQArticleDetail,
} from '../../FAQ/dataloaders/getFAQArticle';
import type { FAQCategoryType } from '../../FAQ/types/outputs/FAQCategory';
import LocaleMap from '../types/enums/LocaleMap';

/**
 * FIXME:
 * This is stupid - it's already defined by data-loader itself but currently
 * it's needed to type resolvers. How to do it better?
 */
export type GraphqlContextType = {|
  // DataLoader<K, V>
  locale: string,
  apiToken: ?string,
  dataLoader: {|
    airline: DataLoader<string, ?Airline>,
    booking: DataLoader<number | string, Booking>,
    bookings: BookingsLoader,
    flight: FlightLoader,
    identity: IdentityDataloader,
    location: LocationLoader,
    locationSuggestions: LocationSuggestionsLoader,
    rates: DataLoader<string, ?number>,
    city: DataLoader<string, City[]>,
    hotel: {
      availabilityByLocation: DataLoader<HotelKey, HotelType[]>,
      availabilityByID: DataLoader<HotelKey, HotelType[]>,
      byID: typeof HotelByID,
      cities: HotelCities,
      room: HotelRoomsLoader,
      roomAvailability: HotelRoomAvailabilityLoader,
      roomBedding: typeof HotelRoomBeddingLoader,
      priceStats: DataLoader<
        { searchParams: HotelKey, boundary: 'MAX' | 'MIN' },
        number,
      >,
    },
    FAQ: DataLoader<FAQArgs, FAQArticleItem[]>,
    FAQCategories: DataLoader<FAQCategoryArgs, FAQCategoryType[]>,
    FAQArticle: DataLoader<FAQCArticleArgs, FAQArticleDetail>,
  |},
  options: OptionsStorage,
  _traceCollector?: Object,
|};

export function createContext(
  token: ?string,
  acceptLanguage: ?string,
): GraphqlContextType {
  const locale =
    acceptLanguage && LocaleMap.hasOwnProperty(acceptLanguage)
      ? acceptLanguage
      : 'en_US';
  const bookings = new BookingsLoader(token);
  const locationSuggestions = new LocationSuggestionsLoader();
  const location = new LocationLoader(locationSuggestions);
  const hotelCities = new HotelCities();

  return {
    locale,
    apiToken: token,
    dataLoader: {
      airline: createAirlineLoader(),
      booking: createBookingLoader(token, bookings),
      bookings: bookings,
      flight: new FlightLoader(location),
      identity: new IdentityDataloader(token),
      location: location,
      locationSuggestions: locationSuggestions,
      rates: createRatesLoader(),
      city: CitiesByID,
      hotel: {
        availabilityByLocation: HotelsAvailability,
        availabilityByID: HotelsAvailability,
        byID: HotelByID,
        cities: hotelCities,
        room: new HotelRoomsLoader(),
        roomAvailability: new HotelRoomAvailabilityLoader(),
        roomBedding: HotelRoomBeddingLoader,
        priceStats: PriceStatsLoader,
      },
      FAQ: createFAQLoader(),
      FAQCategories: createFAQCategoryLoader(),
      FAQArticle: createFAQArticleLoader(),
    },
    options: new OptionsStorage(),
  };
}
