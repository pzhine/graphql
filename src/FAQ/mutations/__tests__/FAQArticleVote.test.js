// @flow

import { toGlobalId } from 'graphql-relay';
import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import config from '../../../../config/application';
import articleMockData from '../../datasets/FAQArticle-39.json';

const goodId = '39';
const globalId = toGlobalId('FAQArticle', goodId);

beforeEach(() => {
  RestApiMock.onGet(config.restApiEndpoint.FAQArticle(goodId)).replyWithData(
    articleMockData,
  );
});

describe('FAQArticleVote mutation', () => {
  it('should vote UP', async () => {
    RestApiMock.onPost(
      config.restApiEndpoint.FAQArticleVote(goodId),
    ).replyWithData({
      message: 'yummy',
    });
    const voteMutation = `
      mutation {
        FAQArticleVote(id:"${globalId}", vote:"up") {
          id
          upvotes
          downvotes
        }
      }
    `;
    expect(await graphql(voteMutation)).toMatchSnapshot();
  });
  it('should vote DOWN', async () => {
    RestApiMock.onPost(
      config.restApiEndpoint.FAQArticleVote(goodId),
    ).replyWithData({
      message: 'yummy',
    });
    const voteMutation = `
      mutation {
        FAQArticleVote(id:"${globalId}", vote:"down") {
          id
          upvotes
          downvotes
        }
      }
    `;
    expect(await graphql(voteMutation)).toMatchSnapshot();
  });
  it('should return WRONG vote', async () => {
    RestApiMock.onPost(
      config.restApiEndpoint.FAQArticleVote(goodId),
    ).replyWithData({
      message: 'invalid post body',
    });
    const voteMutation = `
      mutation {
        FAQArticleVote(id:"${globalId}", vote:"wrong") {
          id
          upvotes
          downvotes
        }
      }
    `;
    expect(await graphql(voteMutation)).toMatchSnapshot();
  });
});
