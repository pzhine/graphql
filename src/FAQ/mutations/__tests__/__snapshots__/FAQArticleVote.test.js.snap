// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAQArticleVote mutation should return WRONG vote 1`] = `
Object {
  "data": Object {
    "FAQArticleVote": null,
  },
  "errors": Array [
    [GraphQLError: Votes can only be 'up' or 'down'],
  ],
}
`;

exports[`FAQArticleVote mutation should vote DOWN 1`] = `
Object {
  "data": Object {
    "FAQArticleVote": Object {
      "downvotes": 64,
      "id": "RkFRQXJ0aWNsZTozOQ==",
      "upvotes": 36,
    },
  },
}
`;

exports[`FAQArticleVote mutation should vote UP 1`] = `
Object {
  "data": Object {
    "FAQArticleVote": Object {
      "downvotes": 64,
      "id": "RkFRQXJ0aWNsZTozOQ==",
      "upvotes": 36,
    },
  },
}
`;
